### ISSUE TITLE
[UX] File system settings: Default download method: do not render as radio if there is only one option (Public local files).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1837]
There's two places where this happens...

The *File system* settings page (`/admin/config/media/file-system`):

![backdrop-file_system_settings-default_download_method_settings-only_public_files_enabled](https://cloud.githubusercontent.com/assets/2423362/15093073/731c2dc8-14bf-11e6-9b1a-6407dbd56f52.png)

...and in the "Upload destination" section of the field settings tab for fields of type "File" in content types:

![backdrop-file_field_settings_tab](https://cloud.githubusercontent.com/assets/2423362/15093091/e0cc6b9e-14bf-11e6-94a5-bac17cd922eb.png)

I think that in general in such cases we should do one of the following:

- hide the entire section entirely (do not provide an options section when there are no options)
- use `#edit-file-default-scheme-public { visibility: hidden; }` (hides the input circle - leaves the label) in order to render the single option as a plain, informational text instead of a radio button (related: #1801):

  ![backdrop-file_system_settings-default_download_method_settings-only_public_files_enabled-input_visibility_hidden](https://cloud.githubusercontent.com/assets/2423362/15093173/cb9f76b4-14c2-11e6-9f3a-73e591f27c60.png)

Alternatively (in this specific case) we could a) communicate the fact that there are actually alternative options but have not been configured/enabled yet and b) provide a connect-the-dots solution for people to enable the additional options.

Here's how this could look in the File system options page:

![backdrop-file_system_settings-default_download_method_settings-private_files_shown_as_disabled_option](https://cloud.githubusercontent.com/assets/2423362/15093270/abd5a576-14c5-11e6-961f-a461430885a5.png)

...and in the file field settings tab:

![backdrop-file_field_settings_tab-private_file_shown_as_disabled_option](https://cloud.githubusercontent.com/assets/2423362/15093326/d20caeb8-14c7-11e6-9e87-6b60be295cfb.png)

The link should ideally open a dialog and enable the option upon submit in an AJAXy way so that the user never leaves the form.

**Comment by klonos** [comment]
...while we're at it: #1838

**Comment by klonos** [comment]
...this is now also in the "Destination" vtab, of the "Manage" tab, when editing a file:

<img width="1198" alt="Screen Shot 2019-09-02 at 4 56 55 am" src="https://user-images.githubusercontent.com/2423362/64086004-6ce31a80-cd3e-11e9-9310-127f9f64bb9a.png">

**Comment by BWPanda** [comment]
I like the option to show the disabled private radio too.

**Comment by klonos** [comment]
I was working on #4585 and kept being annoyed by this 🙄 ...lets get this oldie but goodie UX issue out of the way, shall we?: https://github.com/backdrop/backdrop/pull/3632

**Comment by klonos** [comment]
...I've updated the PR to change this (3rd step when uploading a file from `file/add`):
<img width="45%" height="45%" src="https://user-images.githubusercontent.com/2423362/119347224-e6a20100-bca3-11eb-851e-1c3bb2d2f738.png">

to this:
<img width="70%" height="70%" src="https://user-images.githubusercontent.com/2423362/119347292-00dbdf00-bca4-11eb-866f-79a131ca1f3d.png">

In general, wherever the help text for the public/private directories is shown to content editors (i.e. in `file/add` and `file/%/manage`) I went with something simple:

- Public files
  > Select this option if there is no need for access control for the uploaded file(s).
- Private files
  > Select this option if there is need to restrict access to the uploaded file(s).

Whereas in places where the help text is more likely to be seen by site admins/builders with more technical knowledge (i.e. in `admin/config/media/file-system` and `admin/structure/types/manage/%/fields/%/field-settings`), it is shown as such:

- Public local files served by the webserver
  > The use of public files is more efficient, but does not provide any access control.
- Private local files served by Backdrop
  > Private file storage has significantly more overhead than public files, but allows restricted access to files.

**Comment by olafgrabienski** [comment]
@klonos The sentences for content editors about Public and Private files both start with the same phrase:

> Select this option if there is

This way, it's hard to see the differences between the options. Let's put the differences in the beginning, e.g.:

- Public files
  Deliver uploaded file(s) directly, without access control.

- Private files
  Restrict access to the uploaded file(s).

(I guess, the exact wording needs more work.)

**Comment by indigoxela** [comment]
> I guess, the exact wording needs more work.

Yes, wording will be tricky in this case, because "Public files" and "Private files" look soo similar. Not sure, if there's a smart solution for that.

I like the direction of this issue - which gained momentum recently.

Regarding the plan to add a link to all field settings, that opens the file system settings form via ajax... Personally, I'm not a big fan of these nested workflows. They tend to be confusing and are usually harder to maintain. IMO only providing the text might be less distracting. But that's just my two cents... :grin:

**Comment by klonos** [comment]
> Regarding the plan to add a link to all field settings, that opens the file system settings form via ajax... Personally, I'm not a big fan of these nested workflows.

Yeah, I've created this issue back in 2016, and back then I had a mindset of trying to limit the times people have to pogo-stick through the admin UI in order to complete a single task.

**Comment by indigoxela** [comment]
> trying to limit the times people have to pogo-stick through the admin UI in order to complete a single task.

I see. But the private file system is configured once, and only once. A hint (as text) is nice, in case they're not aware of that option, but links/ajax seem unnecessary here. Again - maybe just my personal opinion. :smiley:

**Comment by klonos** [comment]
> ...but links/ajax seem unnecessary here.

Oh, I thought you were only referring to the AJAX bit, which I might have mentioned in the original issue summary back in 2016, but I did NOT implement in the PR.

But you seem to also be concerned about the "File system setttings" link in `admin/structure/types/manage/%/fields/%/field-settings`:

<img width="70%" alt="Screen Shot 2021-05-28 at 4 12 20 pm" src="https://user-images.githubusercontent.com/2423362/119989198-92fb2480-bfcf-11eb-8fac-032ec2bc9f2d.png">

Do you feel that that should be removed? (there is a check for `user_access('administer site configuration')`, so it should only be shown to site builders/admins - not to content editors).

**Comment by indigoxela** [comment]
> Do you feel that that should be removed?

No, everything's fine with that link, I updated my previous comment. :+1:

**Comment by klonos** [comment]
> UPDATE: I changed my mind, the link is totally OK. It's just ajax that makes no sense for me here. 😜

Yup, that how I feel as well 👍🏼 ...that's how I've implemented this in the current PR. I did that mainly because that would take longer and it would be a bigger change, but also because it is complicated, and we don't have such a pattern elsewhere in core that I can think of.

To be fair though (and why the whole AJAXy thing may be a follow-up - not to this feature here specifically, but as a generic pattern used in the admin UI in Backdrop), if you think about it from a purely UX point of view:
- try to select an upload destination -> see that private is not configured -> open a new tab -> configure it -> save configuration -> return to the previous tab -> refresh the page -> configure to use private files -> save (that's roughly 9 steps)
- try to select an upload destination -> see that private is not configured -> click the AJAXy config thing -> configure it -> save configuration -> options on the form are refreshed automatically -> configure to use private files -> save (that's roughly 7 steps + 1 automated)

So implementing such a "smart" way of "doing things on the spot" is better UX.

**Comment by indigoxela** [comment]
Again: the private file system is configured once, and only once. After it's done, that configure form won't ever be needed again.

But actually that's off topic here, anyway. I got confused by the issue description and am happy that this ajax thing hasn't been implemented.
